### PR TITLE
Read FIXED_LEN_BYTE_ARRAY as binary in parquet reader

### LINKED ISSUE #12590: [FEA] Support reading FIXED_LEN_BYTE_ARRAY as binary in read_parquet
**Is your feature request related to a problem? Please describe.**
Currently, the libcudf parquet reader can only interpret FIXED_LEN_BYTE_ARRAY physical types as decimal. Spark3.4 includes a ([feature](https://issues.apache.org/jira/browse/SPARK-41096)) to read parquet FIXED_LEN_BYTE_ARRAY as binary type (i.e. list of int8 in libcudf). I would like to support reading FIXED_LEN_BYTE_ARRAY as binary in read_parquet

**Describe the solution you'd like**
Add FIXED_LEN_BYTE_ARRAY to list of types that can be read binary (list of int8) in libcudf read_parquet.

**Describe alternatives you've considered**
n/a

**Additional context**
Spark3.4 includes a ([feature](https://issues.apache.org/jira/browse/SPARK-41096)) to read parquet FIXED_LEN_BYTE_ARRAY as binary type (i.e. list of int8 in libcudf). The Spark-RAPIDS team has also opened an [issue](https://github.com/NVIDIA/spark-rapids/issues/7449), to provide a little more context. It seems like [11160](https://github.com/rapidsai/cudf/pull/11160) added support for reading variable-length BYTE_ARRAY, so adding FIXED_LEN_BYTE_ARRAY may not be too complex.

### PR DESCRIPTION
**Pull request by PointKernel**
## Description
Closes #12590 

This PR adds support of reading `FIXED_LEN_BYTE_ARRAY` as lists of `INT8` in the parquet reader.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
Can this be deferred until #13302 is finished (or rejected >.<)? I think the fixed width byte arrays can be processed more efficiently since they'll behave like any other fixed-width type (sizing can be known up front).

**Comment by PointKernel** [comment]
> Can this be deferred until #13302 is finished (or rejected >.<)? I think the fixed width byte arrays can be processed more efficiently since they'll behave like any other fixed-width type (sizing can be known up front).

@etseidl sure, this is still a draft PR and I'm putting it up just for testing.

**Comment by revans2** [comment]
I ran some tests with this and it looks good to me.  Note that I just did some very simple tests though.

**Comment by PointKernel** [comment]
> I ran some tests with this and it looks good to me. Note that I just did some very simple tests though.

Thanks for testing. I have some improvements to apply once #13302 is merged. In the meanwhile, @revans2 please let me know if you find any uncovered issues with the current work.

File: cpp/src/io/parquet/page_data.cu [hunk 1]
```diff
@@ -603,7 +611,7 @@ inline __device__ void gpuOutputString(volatile page_state_s* s,
                                        void* dstv)
 {
   auto [ptr, len] = gpuGetStringData(s, sb, src_pos);
-  if (s->dtype_len == 4) {
+  if (s->dtype_len == 4 and (s->col.data_type & 7) == BYTE_ARRAY) {
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/io/parquet/page_data.cu:44]: Before this PR, only `BYTE_ARRAY` invokes `gpuOutputString` and the input length cannot be 4 in that case. Now with function being potentially invoked by `FIXED_LEN_BYTE_ARRAY` where the length could be 4, this `and` logic is needed.

File: cpp/src/io/parquet/page_data.cu [hunk 2]
```diff
@@ -881,6 +884,12 @@ __global__ void __launch_bounds__(decode_block_size) gpuDecodePageData(
               }
               break;
           }
+        } else if (dtype == FIXED_LEN_BYTE_ARRAY) {
+          gpuOutputString(s,
+                          sb,
+                          val_src_pos,
+                          nesting_info_base[leaf_level_index].data_out +
+                            static_cast<size_t>(dst_pos) * sizeof(string_index_pair));
```
Comment (PointKernel) [inline cpp/src/io/parquet/page_data.cu:892]: Need to update `dtype_len` instead of using `string_index_pair` explicitly.

File: cpp/src/io/parquet/page_string_decode.cu [hunk 3]
```diff
@@ -487,6 +487,36 @@ __device__ size_t totalDictEntriesSize(uint8_t const* data,
...
 
+/**
+ * @brief Compute string size information for fixed len byte array strings.
+ *
+ * @param input_size Size of each fixed length byte array
+ * @param data_size Length of data
+ * @param start_value Do not count values that occur before this index
+ * @param end_value Do not count values that occur after this index
+ */
+__device__ size_t totalFixedLenByteArraySize(int input_size,
```
Comment (etseidl) [inline cpp/src/io/parquet/page_string_decode.cu:498]: couldn't this just return `(end_value - start_value) * dtype_len`?

File: cpp/src/io/parquet/page_string_decode.cu [hunk 4]
```diff
@@ -625,8 +655,12 @@ __global__ void __launch_bounds__(preprocess_block_size) gpuComputePageStringSiz
       break;
     case Encoding::PLAIN:
       dict_size = static_cast<int32_t>(end - data);
-      str_bytes = is_bounds_pg ? totalPlainEntriesSize(data, dict_size, start_value, end_value)
-                               : dict_size - sizeof(int) * (pp->num_input_values - pp->num_nulls);
+      if ((s->col.data_type & 7) == FIXED_LEN_BYTE_ARRAY) {
+        str_bytes = totalFixedLenByteArraySize(s->dtype_len_in, dict_size, start_value, end_value);
```
Comment (etseidl) [inline cpp/src/io/parquet/page_string_decode.cu:659]: or just get rid of `totalFixedLenByteArraySize` and calculate `str_bytes` directly

**Comment by etseidl** [comment]
So will this take `FIXED_LEN_BYTE_ARRAY` data in and produce a string column on output? Would it be better to produce a fixed length `UINT8` array? I'm a little bit worried about getting the column statistics right.

**Comment by PointKernel** [comment]
> So will this take `FIXED_LEN_BYTE_ARRAY` data in and produce a string column on output? Would it be better to produce a fixed length `UINT8` array? I'm a little bit worried about getting the column statistics right.

Right, similar to `BYTE_ARRAY`, `FIXED_LEN_BYTE_ARRAY` will be handled as string columns by default.

> Would it be better to produce a fixed length `UINT8` array? 

This work is based on the assumption that users may switch between string and binary as the eventual output of `FIXED_LEN_BYTE_ARRAY`. Are you suggesting producing binary as default and converting back to strings when requested?

**Comment by etseidl** [comment]
> This work is based on the assumption that users may switch between string and binary as the eventual output of `FIXED_LEN_BYTE_ARRAY`. Are you suggesting producing binary as default and converting back to strings when requested?

The Parquet spec doesn't really allow for `FIXED_LEN_BYTE_ARRAY` to be interpreted as UTF8, so yes, I'd suggest producing binary. That seems to be the assumption behind NVIDIA/spark-rapids#8404 as well.

File: cpp/src/io/parquet/page_data.cu [hunk 5]
```diff
@@ -2027,11 +2035,14 @@ __global__ void __launch_bounds__(decode_block_size) gpuDecodePageData(
     return;
   }
 
+  auto const data_type = s->col.data_type & 7;
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/io/parquet/page_data.cu]: `data_type` is not used that often in other functions. Let me have a further look.

**Comment by PointKernel** [comment]
Convert back to draft since `FIXED_LEN_BYTE_ARRAY` shouldn't be handled as strings.

File: cpp/src/io/parquet/page_string_decode.cu [hunk 6]
```diff
@@ -625,8 +625,12 @@ __global__ void __launch_bounds__(preprocess_block_size) gpuComputePageStringSiz
       break;
```
Comment (etseidl) [inline cpp/src/io/parquet/page_string_decode.cu:625]: I think here too we can short circuit the call to `totalDictEntriesSize` with the same computation as below.

File: cpp/src/io/parquet/page_string_decode.cu [hunk 7]
```diff
@@ -625,8 +625,12 @@ __global__ void __launch_bounds__(preprocess_block_size) gpuComputePageStringSiz
...
     case Encoding::PLAIN:
       dict_size = static_cast<int32_t>(end - data);
-      str_bytes = is_bounds_pg ? totalPlainEntriesSize(data, dict_size, start_value, end_value)
-                               : dict_size - sizeof(int) * (pp->num_input_values - pp->num_nulls);
+      if ((s->col.data_type & 7) == FIXED_LEN_BYTE_ARRAY) {
+        str_bytes = (end_value - start_value) * s->dtype_len_in;
```
Comment (etseidl) [inline cpp/src/io/parquet/page_string_decode.cu:629]: Also forgot that `pp->num_valids` is always equal to `end_value - start_value` so this can be simplified.

File: cpp/src/io/parquet/page_string_decode.cu [hunk 8]
```diff
@@ -625,8 +625,12 @@ __global__ void __launch_bounds__(preprocess_block_size) gpuComputePageStringSiz
...
+      } else {
+        str_bytes = is_bounds_pg ? totalPlainEntriesSize(data, dict_size, start_value, end_value)
+                                 : dict_size - sizeof(int) * (pp->num_input_values - pp->num_nulls);
```
Comment (etseidl) [inline cpp/src/io/parquet/page_string_decode.cu:632]: And here too, `pp->num_valids` is `pp->num_input_values - pp->num_nulls` in the not-bounds-page case.

**Review by etseidl — COMMENTED** [review]
Looking good, just one small suggestion.

File: cpp/src/io/parquet/page_data.cu [hunk 9]
```diff
@@ -774,8 +774,12 @@ __global__ void __launch_bounds__(decode_block_size) gpuDecodePageData(
   if (s->dict_base) {
     out_thread0 = (s->dict_bits > 0) ? 64 : 32;
   } else {
-    out_thread0 =
-      ((s->col.data_type & 7) == BOOLEAN || (s->col.data_type & 7) == BYTE_ARRAY) ? 64 : 32;
+    switch (s->col.data_type) {
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/io/parquet/page_data.cu]: Good catch. I thought the `ColumnChunkDesc::data_type` is also `int8_t` thus taking lower bits with `& 7` was not necessary. Fixed.

File: python/cudf/cudf/tests/data/parquet/fixed_len_byte_array.parquet
[the remark this answers is not in the document]
Comment (GregoryKimball) [reply python/cudf/cudf/tests/data/parquet/fixed_len_byte_array.parquet:1]: @revans2 or @jlowe, would you please create a smaller test file for us?

[the remark this answers is not in the document]
Comment (PointKernel) [reply python/cudf/cudf/tests/data/parquet/fixed_len_byte_array.parquet:1]: @revans2 Updated. Thanks

File: cpp/src/io/parquet/page_data.cu (hunk 1 above)
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/parquet/page_data.cu:44]: maybe a comment along the lines of "// make sure to only hash variable length byte arrays when specified with the output type size"?

File: cpp/src/io/parquet/page_data.cu [hunk 10]
```diff
@@ -41,7 +41,7 @@ inline __device__ void gpuOutputString(volatile page_state_s* s,
                                        void* dstv)
 {
   auto [ptr, len] = gpuGetStringData(s, sb, src_pos);
-  if (s->dtype_len == 4) {
+  if (s->dtype_len == 4 and (s->col.data_type & 7) == BYTE_ARRAY) {
```
Comment (vuule) [inline cpp/src/io/parquet/page_data.cu:44]: not really related to this PR, but we have this `dtype_len == 4` condition as a stand-in for "output hash" in a few places and it really requires detailed knowledge of the code to understand. Nothing actionable, I just don't like it :D

File: cpp/src/io/parquet/parquet_gpu.hpp [hunk 11]
```diff
@@ -432,8 +432,9 @@ struct EncPage {
  */
 constexpr bool is_string_col(ColumnChunkDesc const& chunk)
 {
-  return (chunk.data_type & 7) == BYTE_ARRAY and (chunk.data_type >> 3) != 4 and
-         chunk.converted_type != DECIMAL;
+  return ((chunk.data_type & 7) == BYTE_ARRAY and (chunk.data_type >> 3) != 4 and
```
Comment (vuule) [inline cpp/src/io/parquet/parquet_gpu.hpp:435]: if you don't mind, please break this condition into pieces and give them names. It's getting a bit too hard to follow.
Something like:
```
auto const not_converted_to_decimal = chunk.converted_type != DECIMAL;
auto const non_hashed_byte_array = (chunk.data_type & 7) == BYTE_ARRAY and (chunk.data_type >> 3) != 4
auto const fixed_len_byte_array = (chunk.data_type & 7) == FIXED_LEN_BYTE_ARRAY;
return not_converted_to_decimal and (non_hashed_byte_array or fixed_len_byte_array)
```

**Review by vuule — COMMENTED** [review]
Small change, but tricky! Cool stuff 👍 
Partial review, will continue tomorrow.

File: cpp/src/io/parquet/page_decode.cuh [hunk 12]
```diff
@@ -1147,16 +1155,20 @@ inline __device__ bool setupLocalPageInfo(page_state_s* const s,
...
+            if (dtype_len <= sizeof(int32_t)) {
+              return sizeof(int32_t);
+            } else if (dtype_len <= sizeof(int64_t)) {
+              return sizeof(int64_t);
+            } else {
+              return sizeof(__int128_t);
+            }
+          }();
+        } else {
+          s->dtype_len = sizeof(string_index_pair);
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/page_decode.cuh:1170]: This is another case of the type len indicating the type :(.  Look at the comment around line 1225 in the new file...something about "dtype_len is a lie". The real len should be `sizeof(size_type)`, but then we'd think it's a hash. So for strings (and now fixed_len_byte_array) we just leave it as `sizeof(string_index_pair)`, which it would have had to be before the sweeping string decode changes.

[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/io/parquet/page_decode.cuh:1170]: > Is this the "FIXED_LEN_BYTE_ARRAY but not DECIMAL" case(hard to match block in the diff)?

Yes, we decode data as strings as an intermediate output and eventually the final list will be generated from strings. This is for now the best way to take flat data as input and produce nested output.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/parquet/page_decode.cuh:1170]: Thank you for the explanations, now this makes sense.

File: cpp/src/io/parquet/reader_impl.cpp [hunk 13]
```diff
@@ -382,9 +382,15 @@ table_with_metadata reader::impl::read_chunk_internal(
...
   // Create the final output cudf columns.
   for (size_t i = 0; i < _output_buffers.size(); ++i) {
-    auto const metadata = _reader_column_schema.has_value()
-                            ? std::make_optional<reader_column_schema>((*_reader_column_schema)[i])
-                            : std::nullopt;
+    auto metadata      = _reader_column_schema.has_value()
+                           ? std::make_optional<reader_column_schema>((*_reader_column_schema)[i])
+                           : std::nullopt;
+    auto const& schema = _metadata->get_schema(_output_column_schemas[i]);
+    // FIXED_LEN_BYTE_ARRAY always read as binary
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl.cpp:389]: Current comment is not quite correct because of decimal.
```suggestion
    // FIXED_LEN_BYTE_ARRAY never read as string
```

**Review by vuule — APPROVED** [review]
looks good, just a small suggestion

File: cpp/src/io/parquet/page_data.cu (hunk 5 above)
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/io/parquet/page_data.cu]: Can we use `parquet::Type` directly instead of `uint16_t`?
https://github.com/rapidsai/cudf/blob/41f0caf53662cfde8146647574e705982eb558b1/cpp/src/io/parquet/parquet_gpu.hpp#L266-L267

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/parquet/page_data.cu]: (for some reason) data_type holds the type in the last three bits and length in the rest of the number. Using Type here would make this data member even more misleading.

**Review by hyperbolic2346 — APPROVED** [review]

**Review by revans2 — APPROVED** [review]
Reran my tests and they still all pass

File: python/cudf/cudf/tests/test_parquet.py [hunk 14]
```diff
@@ -2568,6 +2568,15 @@ def test_parquet_reader_binary_decimal(datadir):
     assert_eq(expect, got)
 
 
+def test_parquet_reader_fixed_bin(datadir):
+    fname = datadir / "fixed_len_byte_array.parquet"
+
+    expect = pd.read_parquet(fname)
+    got = cudf.read_parquet(fname).to_pandas()
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_parquet.py:2575]:
```suggestion
    got = cudf.read_parquet(fname)
```

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py:2575]: `assert_eq` will take care of doing a `to_pandas()` call

**Review by galipremsagar — APPROVED** [review]
